extensions/amp-auto-ads/0.1/placement.js

```diff
@@ -0,0 +1,224 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../../../src/log';
+import {LayoutRectDef} from '../../../src/layout-rect';
+import {resourcesForDoc} from '../../../src/resources';
+
+/** @const */
+const TAG = 'amp-auto-ads';
+
+/**
+ * TODO: Specify this via the configuration.
+ * @const
+ */
+const TARGET_AD_HEIGHT_PX = 100;
+
+/** @typedef {{name: string, value: (boolean|number|string)}} */
+export let DataAttributeDef;
+
+/**
+ * @enum {number}
+ */
+export const PlacementState = {
+  UNUSED: 0,
+  RESIZE_FAILED: 1,
+  PLACED: 2,
+};
+
+/**
+ * @enum {number}
+ */
+const Position = {
+  BEFORE: 1,
+  FIRST_CHILD: 2,
+  LAST_CHILD: 3,
+  AFTER: 4,
+};
+
+export class Placement {
+  /**
+   * @param {!Window} win
+   * @param {!Element} anchorElement
+   * @param {!Position} position
+   * @param {!Object} placementObj
+   */
+  constructor(win, anchorElement, position, placementObj) {
+    this.win_ = win;
+
+    this.anchorElement_ = anchorElement;
+
+    this.position_ = position;
+
+    this.placementObj_ = placementObj;
+
+    /** @type {?Element} */
+    this.adElement_ = null;
+
+    /** @type {!PlacementState} */
+    this.state_ = PlacementState.UNUSED;
+  }
+
+  /**
+   * @param {string} type
+   * @param {!Array<!DataAttributeDef>} dataAttributes
+   * @return {!Promise<!PlacementState>}
+   */
+  placeAd(type, dataAttributes) {
+    this.adElement_ = this.createAdElement_(type, dataAttributes);
+    injectElement(this.anchorElement_, this.position_, this.adElement_);
+    return resourcesForDoc(this.adElement_).attemptChangeSize(
```

Comment:
That method doesn't exist on Resources. It's in base-element.js, which means I'd have to get a BaseElement instance for the newly created <AMP-AD>. Not sure how I can do that?